File: Sources/PortWatch/Services/PortInspector.swift
import Foundation

struct PortInspector: Sendable {
    private let runner: ShellCommandRunner

    init(runner: ShellCommandRunner = ShellCommandRunner()) {
        self.runner = runner
    }

    func fetchListeningPorts() throws -> [PortRecord] {
        let lsofResult = try runner.run(
            "/usr/sbin/lsof",
            arguments: ["-iTCP", "-sTCP:LISTEN", "-P", "-n"]
        )

        if lsofResult.exitCode != 0 && lsofResult.stdout.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
            return []
        }

        let baseRecords = parseListeningPorts(from: lsofResult.stdout)
        if baseRecords.isEmpty {
            return []
        }

        let pids = Array(Set(baseRecords.map(\.pid))).sorted()
        let metadata = try fetchProcessMetadata(pids: pids)
        let launchMetadata = try fetchLaunchMetadata(pids: pids)

        return dedupeAndSort(
            baseRecords.map { record in
                let processMetadata = metadata[record.pid]
                let launch = launchMetadata[record.pid]
                let mergedMetadata = merged(record: record, metadata: processMetadata, launch: launch)
                return mergedMetadata
            }
        )
    }

    private func fetchProcessMetadata(pids: [Int]) throws -> [Int: ProcessMetadata] {
        guard !pids.isEmpty else {
            return [:]
        }

        let lsofResult = try runner.run(
            "/usr/sbin/lsof",
            arguments: ["-a", "-d", "cwd,txt", "-Ffnp", "-p", pidList(pids)]
        )
        let psResult = try runner.run(
            "/bin/ps",
            arguments: ["-ww", "-p", pidList(pids), "-o", "pid=", "-o", "lstart=", "-o", "command="]
        )

        var metadata = lsofResult.exitCode == 0 ? parseProcessMetadata(from: lsofResult.stdout) : [:]
        if psResult.exitCode == 0 {
            let psMetadata = parseProcessTimestamps(from: psResult.stdout)
            for (pid, item) in psMetadata {
                var current = metadata[pid] ?? ProcessMetadata()
                current.startedAt = item.startedAt
                current.title = item.title
                current.uptime = item.uptime
                metadata[pid] = current
            }
        }
        return metadata
    }

    private func fetchLaunchMetadata(pids: [Int]) throws -> [Int: LaunchMetadata] {
        guard !pids.isEmpty else {
            return [:]
        }

        let psResult = try runner.run(
            "/bin/ps",
            arguments: ["-axo", "pid=", "-o", "ppid=", "-o", "tty=", "-o", "command="]
        )

        guard psResult.exitCode == 0 else {
            return [:]
        }

        let table = parseProcessTable(from: psResult.stdout)
        var result: [Int: LaunchMetadata] = [:]
        for pid in pids {
            let metadata = inferLaunchMetadata(pid: pid, table: table)
            if metadata.launchedBy != nil || !metadata.launchChain.isEmpty {
                result[pid] = metadata
            }
        }
        return result
    }

    private func merged(
        record: PortRecord,
        metadata: ProcessMetadata?,
        launch: LaunchMetadata?
    ) -> PortRecord {
        let source = resolveSource(for: record, metadata: metadata)
        return PortRecord(
            pid: record.pid,
            command: record.command,
            user: record.user,
            fd: record.fd,
            type: record.type,
            device: record.device,
            sizeOff: record.sizeOff,
            node: record.node,
            endpoint: record.endpoint,
            port: record.port,
            protocolName: record.protocolName,
            state: record.state,
            workingDirectory: metadata?.cwd,
            executablePath: metadata?.executablePath,
            source: source,
            startedAt: metadata?.startedAt,
            uptime: metadata?.uptime,
            launchedBy: launch?.launchedBy,
            launchChain: launch?.launchChain ?? []
        )
    }

    private func resolveSource(for record: PortRecord, metadata: ProcessMetadata?) -> String? {
        guard let metadata else {
            return nil
        }

        if let cwd = metadata.cwd, cwd != "/" {
            return cwd
        }

        if let appName = appName(from: metadata.executablePath ?? metadata.title) {
            return appName
        }

        if let title = metadata.title, title != record.command {
            if title.contains("/") {
                return appName(from: title) ?? basename(of: title)
            }
            return title
        }

        if let executablePath = metadata.executablePath {
            return basename(of: executablePath)
        }

        return metadata.cwd
    }

    private func parseListeningPorts(from output: String) -> [PortRecord] {
        let lines = output.split(separator: "\n", omittingEmptySubsequences: false)
        guard lines.count > 1 else {
            return []
        }

        var records: [PortRecord] = []
        for line in lines.dropFirst() {
            let parts = line.split(
                maxSplits: 8,
                omittingEmptySubsequences: true,
                whereSeparator: \.isWhitespace
            )
            guard parts.count >= 9 else {
                continue
            }

            let command = String(parts[0])
            let pid = Int(parts[1]) ?? -1
            let user = String(parts[2])
            let fd = String(parts[3])
            let type = String(parts[4])
            let device = String(parts[5])
            let sizeOff = String(parts[6])
            let node = String(parts[7])
            let endpoint = String(parts[8])
            let port = parsePort(from: endpoint)

            guard pid > 0, port > 0 else {
                continue
            }

            records.append(
                PortRecord(
                    pid: pid,
                    command: command,
                    user: user,
                    fd: fd,
                    type: type,
                    device: device,
                    sizeOff: sizeOff,
                    node: node,
                    endpoint: endpoint,
                    port: port,
                    protocolName: "TCP",
                    state: "LISTEN",
                    workingDirectory: nil,
                    executablePath: nil,
                    source: nil,
                    startedAt: nil,
                    uptime: nil,
                    launchedBy: nil,
                    launchChain: []
                )
            )
        }

        return records
    }

    private func parsePort(from endpoint: String) -> Int {
        let pattern = /:(\d+)(?:\s+\(LISTEN\))?$/
        guard let match = endpoint.firstMatch(of: pattern) else {
            return 0
        }
        return Int(match.1) ?? 0
    }

    private func parseProcessMetadata(from output: String) -> [Int: ProcessMetadata] {
        var metadata: [Int: ProcessMetadata] = [:]
        var currentPID: Int?
        var currentFD: String?

        for rawLine in output.split(separator: "\n", omittingEmptySubsequences: true) {
            let line = String(rawLine)
            guard let field = line.first else {
                continue
            }
            let value = String(line.dropFirst())

            switch field {
            case "p":
                currentPID = Int(value)
                currentFD = nil
                if let currentPID, metadata[currentPID] == nil {
                    metadata[currentPID] = ProcessMetadata()
                }
            case "f":
                currentFD = value
            case "n":
                guard let currentPID, let currentFD else {
                    continue
                }
                var item = metadata[currentPID] ?? ProcessMetadata()
                if currentFD == "cwd" {
                    item.cwd = value
                } else if currentFD == "txt", item.executablePath == nil {
                    item.executablePath = value
                }
                metadata[currentPID] = item
            default:
                continue
            }
        }

        return metadata
    }

    private func parseProcessTimestamps(from output: String) -> [Int: ProcessMetadata] {
        var metadata: [Int: ProcessMetadata] = [:]
        let formatter = DateFormatter()
        formatter.locale = Locale(identifier: "en_US_POSIX")
        formatter.timeZone = .current
        formatter.dateFormat = "EEE MMM d HH:mm:ss yyyy"

        for rawLine in output.split(separator: "\n", omittingEmptySubsequences: true) {
            let line = rawLine.trimmingCharacters(in: .whitespacesAndNewlines)
            let parts = line.split(whereSeparator: \.isWhitespace)
            guard parts.count >= 6, let pid = Int(parts[0]) else {
                continue
            }

            let startString = parts[1 ... 5].map(String.init).joined(separator: " ")
            let title = parts.count > 6 ? parts[6...].map(String.init).joined(separator: " ") : nil
            let startedAt = formatter.date(from: startString)
            let uptime = startedAt.map { Date().timeIntervalSince($0) }

            metadata[pid] = ProcessMetadata(
                cwd: nil,
                executablePath: nil,
                title: title,
                startedAt: startedAt,
                uptime: uptime
            )
        }

        return metadata
    }

    private func parseProcessTable(from output: String) -> [Int: ProcessTableEntry] {
        var table: [Int: ProcessTableEntry] = [:]

        for rawLine in output.split(separator: "\n", omittingEmptySubsequences: true) {
            let parts = rawLine.split(
                maxSplits: 3,
                omittingEmptySubsequences: true,
                whereSeparator: \.isWhitespace
            )
            guard parts.count >= 4,
                  let pid = Int(parts[0]),
                  let ppid = Int(parts[1])
            else {
                continue
            }

            table[pid] = ProcessTableEntry(
                pid: pid,
                ppid: ppid,
                tty: String(parts[2]),
                command: String(parts[3])
            )
        }

        return table
    }

    private func inferLaunchMetadata(pid: Int, table: [Int: ProcessTableEntry]) -> LaunchMetadata {
        let chain = processChain(for: pid, table: table)
        let visibleChain = chain.compactMap { entry -> VisibleProcess? in
            let name = processDisplayName(for: entry)
            guard !name.isEmpty, !isSystemLaunchProcess(name) else {
                return nil
            }
            return VisibleProcess(name: name, appName: appName(from: entry.command))
        }

        let launchChain = dedupeAdjacent(visibleChain.map(\.name))
        let launchedBy = visibleChain.first(where: { $0.appName != nil })?.appName ?? launchChain.first

        return LaunchMetadata(launchedBy: launchedBy, launchChain: launchChain)
    }

    private func processChain(for pid: Int, table: [Int: ProcessTableEntry]) -> [ProcessTableEntry] {
        var chain: [ProcessTableEntry] = []
        var seen = Set<Int>()
        var currentPID: Int? = pid

        while let pid = currentPID, !seen.contains(pid), seen.count < 32 {
            seen.insert(pid)
            guard let entry = table[pid] else {
                break
            }
            chain.insert(entry, at: 0)
            if entry.ppid <= 0 {
                break
            }
            currentPID = entry.ppid
        }

        return chain
    }

    private func processDisplayName(for entry: ProcessTableEntry) -> String {
        if let appName = appName(from: entry.command) {
            return appName
        }

        let tokens = entry.command.split(whereSeparator: \.isWhitespace).map(String.init)
        guard let executable = tokens.first else {
            return entry.command
        }

        let baseName = basename(of: executable.trimmingCharacters(in: CharacterSet(charactersIn: "-")))
        if ["node", "bun", "python", "python3"].contains(baseName),
           tokens.count > 1,
           !tokens[1].hasPrefix("-")
        {
            return stripScriptExtension(basename(of: tokens[1]))
        }

        return stripScriptExtension(baseName)
    }

    private func dedupeAndSort(_ records: [PortRecord]) -> [PortRecord] {
        var byID: [String: PortRecord] = [:]
        for record in records {
            byID[record.id] = record
        }

        return byID.values.sorted {
            if $0.port == $1.port {
                if $0.command == $1.command {
                    return $0.pid < $1.pid
                }
                return $0.command.localizedCaseInsensitiveCompare($1.command) == .orderedAscending
            }
            return $0.port < $1.port
        }
    }

    private func pidList(_ pids: [Int]) -> String {
        pids.map(String.init).joined(separator: ",")
    }

    private func basename(of path: String) -> String {
        let normalized = path.replacingOccurrences(of: "\\", with: "/").trimmingCharacters(in: CharacterSet(charactersIn: "/"))
        return normalized.split(separator: "/").last.map(String.init) ?? path
    }

    private func stripScriptExtension(_ value: String) -> String {
        value.replacingOccurrences(
            of: #"\.(cjs|js|mjs|ts|tsx)$"#,
            with: "",
            options: .regularExpression
        )
    }

    private func appName(from path: String?) -> String? {
        guard let path else {
            return nil
        }

        let pattern = #"(?:^|/)([^/]+)\.app(?:/|$)"#
        guard let range = path.range(of: pattern, options: .regularExpression) else {
            return nil
        }

        let match = String(path[range])
        return match
            .split(separator: "/")
            .last?
            .replacingOccurrences(of: ".app", with: "")
    }

    private func isSystemLaunchProcess(_ name: String) -> Bool {
        name == "launchd" || name == "login"
    }

    private func dedupeAdjacent(_ values: [String]) -> [String] {
        var deduped: [String] = []
        for value in values where deduped.last != value {
            deduped.append(value)
        }
        return deduped
    }
}

private struct ProcessMetadata: Sendable {
    var cwd: String?
    var executablePath: String?
    var title: String?
    var startedAt: Date?
    var uptime: TimeInterval?
}

private struct ProcessTableEntry: Sendable {
    let pid: Int
    let ppid: Int
    let tty: String
    let command: String
}

private struct LaunchMetadata: Sendable {
    let launchedBy: String?
    let launchChain: [String]
}

private struct VisibleProcess: Sendable {
    let name: String
    let appName: String?
}

File: src/electron/services/port-inspector.ts
import { platform } from 'node:os'
import type { PortRecord } from '../types'
import { runCommand } from './shell'
import { parseWindowsListeningPortsJson } from './windows-port-parser'

interface ProcessMetadata {
  command?: string
  cwd?: string
  executablePath?: string
  startedAt?: string
  uptime?: number
  parentPid?: number
  launchedBy?: string
  launchChain?: string[]
}

interface ProcessTableItem {
  pid: number
  ppid?: number
  command: string
}

export async function fetchListeningPorts(): Promise<PortRecord[]> {
  if (platform() === 'win32') {
    return fetchWindowsListeningPorts()
  }
  return fetchMacListeningPorts()
}

async function fetchMacListeningPorts(): Promise<PortRecord[]> {
  const lsof = await runCommand('/usr/sbin/lsof', ['-iTCP', '-sTCP:LISTEN', '-P', '-n'])
  if (lsof.exitCode !== 0 && lsof.stdout.trim().length === 0) {
    return []
  }

  const baseRecords = parseMacLsof(lsof.stdout)
  const pids = [...new Set(baseRecords.map((record) => record.pid))]
  const [metadata, launchMetadata] = await Promise.all([
    fetchMacProcessMetadata(pids),
    fetchMacLaunchMetadata(pids)
  ])

  return sortAndDedupe(baseRecords.map((record) => {
    const item = metadata.get(record.pid)
    const launch = launchMetadata.get(record.pid)
    const source = resolveSource(record, item)

    return {
      ...record,
      command: item?.command ?? record.command,
      workingDirectory: item?.cwd,
      executablePath: item?.executablePath,
      source,
      startedAt: item?.startedAt,
      uptime: item?.uptime,
      launchedBy: launch?.launchedBy,
      launchChain: launch?.launchChain ?? []
    }
  }))
}

function parseMacLsof(output: string): PortRecord[] {
  const [, ...lines] = output.split(/\r?\n/)
  return lines.flatMap((line) => {
    const parts = line.trim().split(/\s+/, 9)
    if (parts.length < 9) return []

    const [command, rawPid, user, , , , , , endpoint] = parts
    const pid = Number(rawPid)
    const port = parsePort(endpoint)
    if (!Number.isFinite(pid) || !port) return []

    return [{
      id: `${pid}:${port}:${endpoint}`,
      pid,
      command,
      user,
      endpoint,
      address: parseAddress(endpoint),
      port,
      protocolName: 'TCP',
      state: 'LISTEN',
      launchChain: []
    }]
  })
}

async function fetchMacProcessMetadata(pids: number[]): Promise<Map<number, ProcessMetadata>> {
  const metadata = new Map<number, ProcessMetadata>()
  if (pids.length === 0) return metadata

  const pidList = pids.join(',')
  const [lsof, ps] = await Promise.all([
    runCommand('/usr/sbin/lsof', ['-a', '-d', 'cwd,txt', '-Ffnp', '-p', pidList]),
    runCommand('/bin/ps', ['-ww', '-p', pidList, '-o', 'pid=', '-o', 'lstart=', '-o', 'command='])
  ])

  if (lsof.exitCode === 0) {
    let currentPid: number | undefined
    let currentFd: string | undefined
    for (const line of lsof.stdout.split(/\r?\n/)) {
      const field = line[0]
      const value = line.slice(1)
      if (field === 'p') {
        currentPid = Number(value)
        if (!metadata.has(currentPid)) metadata.set(currentPid, {})
        currentFd = undefined
      } else if (field === 'f') {
        currentFd = value
      } else if (field === 'n' && currentPid) {
        const item = metadata.get(currentPid) ?? {}
        if (currentFd === 'cwd') item.cwd = value
        if (currentFd === 'txt' && !item.executablePath) item.executablePath = value
        metadata.set(currentPid, item)
      }
    }
  }

  if (ps.exitCode === 0) {
    const now = Date.now()
    for (const line of ps.stdout.split(/\r?\n/)) {
      const match = line.match(/^\s*(\d+)\s+([A-Z][a-z]{2}\s+[A-Z][a-z]{2}\s+\d+\s+\d{2}:\d{2}:\d{2}\s+\d{4})\s+(.+)$/)
      if (!match) continue
      const pid = Number(match[1])
      const started = new Date(match[2])
      const item = metadata.get(pid) ?? {}
      item.startedAt = Number.isNaN(started.getTime()) ? undefined : started.toISOString()
      item.uptime = item.startedAt ? Math.max(0, (now - started.getTime()) / 1000) : undefined
      item.command = basename(match[3].split(/\s+/)[0])
      metadata.set(pid, item)
    }
  }

  return metadata
}

async function fetchMacLaunchMetadata(pids: number[]): Promise<Map<number, ProcessMetadata>> {
  const result = new Map<number, ProcessMetadata>()
  if (pids.length === 0) return result

  const ps = await runCommand('/bin/ps', ['-axo', 'pid=', '-o', 'ppid=', '-o', 'command='])
  if (ps.exitCode !== 0) return result

  const table = parseProcessTable(ps.stdout)
  for (const pid of pids) {
    const chain: string[] = []
    let current = table.get(pid)
    const seen = new Set<number>()
    while (current?.ppid && !seen.has(current.ppid) && chain.length < 6) {
      seen.add(current.pid)
      const parent = table.get(current.ppid)
      if (!parent) break
      const name = basename(parent.command.split(/\s+/)[0])
      if (name && name !== 'launchd') chain.unshift(name)
      current = parent
    }
    if (chain.length > 0) {
      result.set(pid, { launchedBy: chain[chain.length - 1], launchChain: chain })
    }
  }

  return result
}

function parseProcessTable(output: string): Map<number, ProcessTableItem> {
  const table = new Map<number, ProcessTableItem>()
  for (const line of output.split(/\r?\n/)) {
    const match = line.match(/^\s*(\d+)\s+(\d+)\s+(.+)$/)
    if (!match) continue
    const pid = Number(match[1])
    table.set(pid, { pid, ppid: Number(match[2]), command: match[3] })
  }
  return table
}

async function fetchWindowsListeningPorts(): Promise<PortRecord[]> {
  const script = `
$connections = @(Get-NetTCPConnection -State Listen | Select-Object LocalAddress,LocalPort,OwningProcess,State)
$pids = @($connections | Select-Object -ExpandProperty OwningProcess -Unique)
$processes = @(Get-CimInstance Win32_Process | Where-Object { $pids -contains $_.ProcessId } | Select-Object ProcessId,Name,CommandLine,ExecutablePath,ParentProcessId,CreationDate)
[PSCustomObject]@{ Connections = $connections; Processes = $processes } | ConvertTo-Json -Depth 4
`
  const result = await runCommand('powershell.exe', ['-NoProfile', '-ExecutionPolicy', 'Bypass', '-Command', script], 15000)
  if (result.exitCode !== 0 || result.stdout.trim().length === 0) return []

  return parseWindowsListeningPortsJson(result.stdout)
}

function parsePort(endpoint: string): number {
  const match = endpoint.match(/:(\d+)(?:\s+\(LISTEN\))?$/)
  return match ? Number(match[1]) : 0
}

function parseAddress(endpoint: string): string {
  return endpoint.replace(/:(\d+)(?:\s+\(LISTEN\))?$/, '')
}

function resolveSource(record: PortRecord, metadata?: ProcessMetadata): string | undefined {
  if (!metadata) return undefined
  if (metadata.cwd && metadata.cwd !== '/') return metadata.cwd
  if (metadata.executablePath) return appName(metadata.executablePath) ?? basename(metadata.executablePath)
  return metadata.command ?? record.command
}

function appName(value: string): string | undefined {
  const match = value.match(/\/([^/]+\.app)\//)
  return match?.[1]?.replace(/\.app$/, '')
}

function basename(value: string): string {
  return value.split(/[\\/]/).filter(Boolean).pop() ?? value
}

function sortAndDedupe(records: PortRecord[]): PortRecord[] {
  const byId = new Map(records.map((record) => [record.id, record]))
  return [...byId.values()].sort((a, b) => a.port - b.port || a.pid - b.pid)
}
